cajita/src/Cajita_SparseIndexSpace.hpp

```diff
@@ -0,0 +1,817 @@
+  \tparam Key Type of the tile/cell hash key
+  \tparam Value Type of the tile/cell No.
+*/
+template <typename ExecutionSpace, int N, unsigned long long CBits,
+          unsigned long long CNumPerDim, unsigned long long CNumPerTile,
+          HashTypes Hash, typename Key, typename Value>
+class BlockIndexSpace
+{
+  public:
+    //! Number of bits (per dimension) needed to index the cells inside a tile
+    static constexpr unsigned long long cell_bits_per_tile_dim = CBits;
+    //! Number of cells (per dimension) inside each tile
+    static constexpr unsigned long long cell_num_per_tile_dim = CNumPerDim;
+    //! Number of cells (total) inside each tile
+    static constexpr unsigned long long cell_num_per_tile = CNumPerTile;
+    //! Types
+    using key_type = Key;                        // tile hash key
+    using value_type = Value;                    // tile No.
+    static constexpr HashTypes hash_type = Hash; // hash table
+    using bis_Type = BlockIndexSpace<ExecutionSpace, N, cell_bits_per_tile_dim,
+                                     cell_num_per_tile_dim, cell_num_per_tile,
+                                     hash_type, key_type, value_type>; // itself
+
+    /*!
+      \brief Constructor
+      \param size_x The size of the block (MPI rank) in dim-x (Unit: tile)
+      \param size_y The size of the block (MPI rank) in dim-y (Unit: tile)
+      \param size_z The size of the block (MPI rank) in dim-z (Unit: tile)
+      \param capacity Expected capacity of the allocator to store the tiles
+      \param rehash_factor The factor multiplies to the capacity for resizing,
+      when tile nums exceed the capcity
+    */
+    BlockIndexSpace( const int size_x, const int size_y, const int size_z,
+                     const value_type capacity, const float rehash_factor )
+        : _tile_table_info( "hash_table_info" )
+        , _tile_inserted_label( "insert_label", size_x * size_y * size_z )
+        , _tile_table( size_x * size_y * size_z )
+        , _op_ijk2key( size_x, size_y, size_z )
+        , _op_key2ijk( size_x, size_y, size_z )
+    {
+        // hash table related init
+        _tile_table.clear();
+        auto tile_table_info_mirror =
+            Kokkos::create_mirror_view( Kokkos::HostSpace(), _tile_table_info );
+        tile_table_info_mirror( 0 ) = 0;
+        tile_table_info_mirror( 1 ) = capacity;
+        Kokkos::deep_copy( _tile_table_info, tile_table_info_mirror );
+
+        // size related init
+        _block_size[0] = size_x;
+        _block_size[1] = size_y;
+        _block_size[2] = size_z;
+        int total_size = size_x * size_y * size_z;
+        _rehash_coeff[0] = rehash_factor;
+
+        // insert status related init
+        auto tile_inserted_label_mirror = Kokkos::create_mirror_view(
+            Kokkos::HostSpace(), _tile_inserted_label );
+        for ( int i = 0; i < total_size; i++ )
+        {
+            tile_inserted_label_mirror( i ) = 0;
```

Comment:
for this one you can just do `Kokkos::deep_copy( _tile_inserted_label, 0 )` instead of the explicit mirror creation, loop, and copy.